core/src/test/java/org/apache/iceberg/TestManifestReader.java

```diff
@@ -142,9 +147,61 @@ public void testDeleteFilePositions() throws IOException {
       long expectedPos = 0L;
       for (DeleteFile file : reader) {
         Assert.assertEquals("Position should match", (Long) expectedPos, file.pos());
-        Assert.assertEquals("Position from field index should match", expectedPos, ((BaseFile) file).get(17));
+        Assert.assertEquals("Position from field index should match", expectedPos, ((BaseFile) file).get(18));
         expectedPos += 1;
       }
     }
   }
+
+  @Test
+  public void testManifestCompatible() throws IOException {
+    Schema schema = new Schema(
+        required(1, "id", Types.LongType.get()),
+        required(2, "timestamp", Types.TimestampType.withZone()),
+        required(3, "category", Types.StringType.get()),
+        required(4, "data", Types.StringType.get()),
+        required(5, "double", Types.DoubleType.get()));
+
+    PartitionSpec spec = PartitionSpec.builderFor(schema)
+        .identity("category")
+        .hour("timestamp")
+        .bucket("id", 16)
+        .build();
+    PartitionData partition = DataFiles.data(spec, "category=cheesy/timestamp_hour=10/id_bucket=3");
+
+    String path = "s3://bucket/table/category=cheesy/timestamp_hour=10/id_bucket=3/file.avro";
+    long fileSize = 150972L;
+    FileFormat format = FileFormat.AVRO;
+
+    Metrics metrics = new Metrics(
+        1587L,
+        ImmutableMap.of(1, 15L, 2, 122L, 3, 4021L, 4, 9411L, 5, 15L), // sizes
+        ImmutableMap.of(1, 100L, 2, 100L, 3, 100L, 4, 100L, 5, 100L),  // value counts
+        ImmutableMap.of(1, 0L, 2, 0L, 3, 0L, 4, 0L, 5, 0L), // null value counts
+        ImmutableMap.of(5, 10L), // nan value counts
+        ImmutableMap.of(1, Conversions.toByteBuffer(Types.IntegerType.get(), 1)),  // lower bounds
+        ImmutableMap.of(1, Conversions.toByteBuffer(Types.IntegerType.get(), 1))); // upper bounds
+    Integer sortOrderId = 2;
+
+    String fileName = String.format("OldManifestFileV%s.avro", formatVersion);
```

Comment:
The `OldManifestFileV1.avro/OldManifestFileV2.avro/` is the previous DataFile(the file instance is https://github.com/apache/iceberg/blob/master/core/src/test/java/org/apache/iceberg/TestManifestWriterVersions.java#L74). Hi, @szehon-ho I think this test covered your concern.

Reply:
This is really great to have this test.  I was initially thinking that we want to have a TestV1Writer / TestV2Writer that writes all the optional fields as null, instead of checking in the old version avro file.  Is that possible?   I can also take a look myself if that is possible.

Reply:
I am not sure I have fully understood your ways. In this pr, we not only add a new field to the `DataFile` and also changed the return StructType of `DataFile.getType`. Shouldn't we need to customize DataFile/V1Metadata/V2Metadata to write the DataFile with old data spec?

Reply:
Yea I was thinking to make a Test version of those writers..  not sure if its possible.  Anyway, I guess this works too, only issue is wont be too debuggable if something goes wrong.